scripts/test_solidity_lsp.py

```diff
@@ -0,0 +1,379 @@
+#!/usr/bin/env python3.8
+
+import argparse
+import json
+import os
+import subprocess
+
+from typing import List, Union, Any
+from deepdiff import DeepDiff
+
+# {{{ JsonRpcProcess
+class MyEncoder(json.JSONEncoder):
+    """
+    Encodes an object in JSON
+    """
+    def default(self, o): # pylint: disable=E0202
+        return o.__dict__
+
+class JsonRpcProcess:
+    exe_path: str
+    exe_args: List[str]
+    process: subprocess.Popen
+    trace_io: bool
+
+    def __init__(self, exe_path: str, exe_args: List[str], trace_io: bool = False):
+        self.exe_path = exe_path
+        self.exe_args = exe_args
+        self.trace_io = trace_io
+
+    def __enter__(self):
+        self.process = subprocess.Popen(
+            [self.exe_path, *self.exe_args],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE
+        )
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback) -> None:
+        self.process.kill()
+        self.process.wait(timeout=2.0)
+
+    def trace(self, topic: str, message: str) -> None:
+        if self.trace_io:
+            print(f"{SGR_TRACE}{topic}:{SGR_RESET} {message}")
+
+    def receive_message(self) -> Any:
+        # `, timeout: float = 2.0`
+        # Note, we should make use of timeout to avoid infinite blocking if nothing is received.
+        LEN_HEADER = "Content-Length: "
+        TYPE_HEADER = "Content-Type: "
+        if self.process.stdout == None:
+            return None
+        message_size = None
+        while True:
+            #read header
+            line = self.process.stdout.readline()
+            if not line:
+                # server quit
+                return None
+            line = line.decode("utf-8")
+            if not line.endswith("\r\n"):
+                raise RuntimeError("Bad header: missing newline")
+            # remove the "\r\n"
+            line = line[:-2]
+            if line == "":
+                break # done with the headers
+            if line.startswith(LEN_HEADER):
+                line = line[len(LEN_HEADER):]
+                if not line.isdigit():
+                    raise RuntimeError("Bad header: size is not int")
+                message_size = int(line)
+            elif line.startswith(TYPE_HEADER):
+                # nothing todo with type for now.
+                pass
+            else:
+                raise RuntimeError("Bad header: unknown header")
+        if not message_size:
+            raise RuntimeError("Bad header: missing size")
```

Comment:
It will raise an error when `message_size` is 0. I think that's not intentional?
```suggestion
        if message_size is None:
            raise RuntimeError("Bad header: missing size")
```
This is the reason why I think this common convention of using just `if var:` for non-bool variables in Python is a bad idea. I never use it myself.